### ISSUE TITLE
Slow & incomplete 3D-milling - bug in surface 3d-mill component?

### ISSUE THREAD

**Issue by Siemenc** [fellesverkstedet/Bark-beetle-parametric-toolpaths#52]
I get slow and incomplete 3D-milling toolpaths when calculating a fairly simple surface.
![image](https://user-images.githubusercontent.com/10272127/61635082-0aa0ef80-ac93-11e9-8c4e-7824e1d8e55d.png)
![image](https://user-images.githubusercontent.com/10272127/61635117-1f7d8300-ac93-11e9-8566-6e845a938417.png)

When investigating I think it might have something to do with the difference in tree branches inside the surface 3D mill component. Looking further into this.
![image](https://user-images.githubusercontent.com/10272127/61635319-987cda80-ac93-11e9-8ce4-f036c085a4ff.png)

**Comment by Siemenc** [comment]
My assumptions were correct, I think. I went for the same surface and same settings from this:
![image](https://user-images.githubusercontent.com/10272127/61635783-98c9a580-ac94-11e9-82bc-e6f7f3d7d1b7.png)
![image](https://user-images.githubusercontent.com/10272127/61635761-88b1c600-ac94-11e9-89a6-1fc636a0bda2.png)

To this:
![image](https://user-images.githubusercontent.com/10272127/61635834-b4cd4700-ac94-11e9-92c8-09899cb0eb17.png)
![image](https://user-images.githubusercontent.com/10272127/61635879-c9114400-ac94-11e9-8cd3-20aa543f7e71.png)

And this toolpath output:
![image](https://user-images.githubusercontent.com/10272127/61636003-083f9500-ac95-11e9-8b72-3aa571175caa.png)

I'll be doing quite some 3D-milling in the coming weeks so I'll get to crash test this and see if any other bugs pop up by fixing this.

**Comment by JensDyvik** [comment]
This is great bug to iron out. Thank you! I can not remember having problems with simple surfaces before. But lately I have had many data tree problems in rhino6 in definitions made in GH in rhino5 *(especially 3D milling from meshes and tabs on 3D objects)*. Maybe some ways of handling data trees and paths have changed in new GH?

Also important, we need to find a better way to publish the latest version of a complete bark beetle definition *(or example of use of components)*. For now, could you update this file with your fixes? https://github.com/fellesverkstedet/Bark-beetle-parametric-toolpaths/tree/master/Examples%20-%20Stable%20release

**Comment by Siemenc** [comment]
My findings might have been based upon a previous version. When I opened the last Rhino6 file and tried getting the same issue I couldn't repeat the behavior. Will do some more testing before fixing and committing.

**Comment by JensDyvik** [comment]
I have been making some improvements to 3D milling in "Bark beetle 1.02 beta 5 - CNC milling - Rhino6.gh" published here: https://github.com/fellesverkstedet/Bark-beetle-parametric-toolpaths/tree/master/Development%20files

The improvements are:

- Stepover distance is much more consistent, no more need to rebuild curves in some circumstances. The stepover is now based on dividing a sample isoccurve of a 90 deg direction in fixed lengths. The sampled isocurve is the longest of out 20 _(adjustable inside cluster)_.
- I added the option to disable collision checking _(this makes 3D milling calculation at least 10 times faster, but at the cost of risking to mill away too much material at corners and overhangs)_. The option is in the bottom of "3D roughing settings", but also impacts 3D finishing.
- Toolpaths are now more reliable in starting at the highest point.
- Weaving seams are a little tidier thanks to better sample point making for collision detection
- For roughing I turned off "Include surface edges" as default. This was seldom needed, but created messy and confusing toolpaths _(the option was added in order to reduce the risk of roughing pass leaving too much stock behind)_. 

Testing and feedback is very welcome and appreciated!

![old isocurve dist](https://user-images.githubusercontent.com/6598148/113616203-17c36500-9655-11eb-9454-3fe57882bee9.PNG)
_Distrubtion of isocurve based toolpaths the old way_

![new isocurve dist](https://user-images.githubusercontent.com/6598148/113616223-1c881900-9655-11eb-9e53-e65cdc5662c5.PNG)
__Distrubtion of isocurve based toolpaths with the new approach_

![example no collsion detection](https://user-images.githubusercontent.com/6598148/113616439-5c4f0080-9655-11eb-9b90-e803457b02ac.PNG)
_Example of corners getting milled into with "Collsion detection" turned off_

![example with collsion detection](https://user-images.githubusercontent.com/6598148/113616462-63760e80-9655-11eb-892f-f56e05da4563.PNG)
_With "Collision detection" on everything is nice and clean, but takes about 10 times longer to calculate_

**Comment by Siemenc** [comment]
Nice progress, looks good! Do I get it correct that if you have single face surfaces without undercuts, you can always leave collision detection turned off without having to worry?

Also, just to be sure: In the current beta versions: are they a continuation of each other or does each different beta version contain an improved part which would need to be combined before becoming a new version?

**Comment by JensDyvik** [comment]
Yes, in most single surface 3D milling scenarios it would be safe to leave collision detection off. The exemption would be a curving surface having a quite sharp internal corner _(radius of corner smaller than radius of the milling bit)_ or overhangs like you mention. All flat single surfaces should be very safe to calculate without collision detection.

The betas are a continuation off eachother. Otherwise I think it would be too much work to merge different development _(since there is no git merge/diff support for GH files)_. So I am thinking if one of us introduces a new bug in a beta, we can copy and paste a better solution from an older file. I think we should all try to do new work in _(or paste into)_  the latest beta file for best collaboration. What do you think?
